include/copy.hpp

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include "filesystem"
+#include <fstream>
+#include <vector>
+#include <iostream>
+#include <thread>
+#include <mutex>
+
+static int COPY_BUFFER_SIZE = 1024;
+
+class Copy {
+public:
+    Copy(std::filesystem::path source, std::filesystem::path destination){
+        _source = source;
+        _destination = destination;
+    }
+
+    void run_one_thread(){
+        std::ofstream fout(_destination, std::ifstream::binary);
+        std::ifstream fin(_source, std::ifstream::binary);
+        char* buffer = new char[COPY_BUFFER_SIZE];
+
+        try {
+            while(!fin.eof()) {
+                fin.read(buffer, COPY_BUFFER_SIZE);
+                std::streamsize s=fin.gcount();
+                fout.write(buffer, s);
+            }
+        }
+        catch(...) {
+            delete[] buffer;
+            try {
+                fout.close();
+                fin.close();
+            }
+            catch (...){
+                std::cerr << "Error while closing working files"<< std::endl;
+            }
+            throw ;
+        }
+        delete[] buffer;
+        fout.close();
+        fin.close();
+    }
+
+    void parallel_copy(){
+        char* buffer = new char[COPY_BUFFER_SIZE];
+
+        std::thread reader(&Copy::reader, this, buffer);
+        std::thread writer(&Copy::writer, this, buffer);
+
+        reader.join();
+        writer.join();
+
+        delete[] buffer;
+    }
+
+    void writer(char* buffer) {
+        std::cout << "start writer worker"<< std::endl;
+        std::ofstream fout(_destination, std::ifstream::binary);
+
+        do {
+            std::unique_lock lk(_m);
+
+            if (actual_buffer_size > 0) {
+                fout.write(buffer, actual_buffer_size);
+            }
+            lk.unlock();
+        } while (!reading_done);
+
+        fout.close();
+    }
+
+    void reader(char* buffer) {
```

Comment:
The multithreading approach actually would be slower than single threaded, if you try to benchmark them.
Questions:
- why so?
- do you have any idea how to improve this?